src/CommonRuntime/Caching.fs

```diff
@@ -35,16 +36,19 @@ let createInMemoryCache expiration =
 open System.Collections.Concurrent
 
 /// Creates a cache that uses in-memory collection
-let createInMemoryCache expiration = 
+let createInMemoryCache (expiration:TimeSpan) = 
   let dict = new ConcurrentDictionary<_, _>()
   { new ICache<_> with
       member __.Set(key, value) = 
-        dict.[key] <- (value, DateTime.UtcNow)
+        dict.AddOrUpdate(key, (value, DateTime.UtcNow), (fun _ _ -> value, DateTime.UtcNow)) |> ignore
```

Comment:
To make it more clear:
Executing this in 10 threads within 10secs will take gigabyte of memory even when cache is ConcurrentDictionary:

```
match cache.TryRetrieve typeName with 
| Some x -> x 
| None -> 
   let x = long_10sec_call_taking_100MB_of_temp_memory()
   cache <- x
   x
```

Because cache is accessed twice in the code and ConcurrentDictionary doesn't have lock over multiple calls.
This is exactly the same functionality, but taking just 100MB even when done in multiple threads:

```
cache.GetOrAdd(typeName, fun typeName -> 
   long_10sec_call_taking_100MB_of_temp_memory())
```

For that reason I think current ICache<> is a bad idea.